Duckling/Numeral/DE/NumParser.hs

```diff
@@ -0,0 +1,154 @@
+module Duckling.Numeral.DE.NumParser (parseNumeral) where
+
+import           Control.Applicative
+import           Data.Char
+import           Data.List
+
+newtype Parser a = Parser { runParser :: String -> Maybe (a, String) }
+
+char :: Char -> Parser Char
+char c = Parser p
+  where
+    p [] = Nothing
+    p (x:xs)
+      | x == c    = Just (x, xs)
+      | otherwise = Nothing
+
+inParser f = Parser . f . runParser
+
+first :: (a -> b) -> (a,c) -> (b,c)
+first f (x,y) = (f x, y)
+
+instance Functor Parser where
+  fmap = inParser . fmap . fmap . first
+
+instance Applicative Parser where
+  pure a = Parser (\s -> Just (a, s))
+  (Parser fp) <*> xp = Parser $ \s ->
+    case fp s of
+      Nothing     -> Nothing
+      Just (f,s') -> runParser (f <$> xp) s'
+
+instance Alternative Parser where
+  empty = Parser (const Nothing)
+  Parser p1 <|> Parser p2 = Parser $ liftA2 (<|>) p1 p2
+
+type NumParser = Parser Integer
+
+(.+.) :: NumParser -> NumParser -> NumParser
+(.+.) = (<*>) . ((+) <$>)
+
+(.*.) :: NumParser -> NumParser -> NumParser
+(.*.) = (<*>) . ((*) <$>)
+
+(?) :: NumParser -> NumParser
+(?) = (<|> Parser p) where p s = Just (0, s)
```

Comment:
this operator is questionable. if we fail to parse, why not just fail, eg with MonadFail or something. You can either change the definition of Parser to run in `Either String` or keep it as `Maybe` and just return `Left err` or `Nothing`. `Either` would be better for debugging, and you could just discard the error and turn into a Maybe when this module is consumed (since it wouldn't be useful in the rest of duckling anyway).

Reply:
Unfortunately, I cannot follow here. Let me explain my intent with ```(?)```, now renamed to ```opt```: ```opt``` makes a parser optional. To illustrate, if ```hundred``` is a parser that maps "hunderteins" to ```Just (100, "eins")``` and ```one``` is a parser that maps "eins" to ```Just (1, "")```, then ```hundert .+. opt and .+. eins``` is a parser that maps "hunderteins" as well as "hundertundeins" to ```Just (101, "")```.  Unfortunately, I cannot speak to the deeper points you're making because I lack insights into the architecture of duckling. So I'd like to suggest to keep Parser as is for the moment, and come back to your suggestion when extending NumParser to German compounds like "zweitausendeinhundertjährig" 'two thousand one hundred years old' and to Italian numerals (see below for further comments).

Reply:
I think now I understand the suggestion by @chessai and how to implement it. Would you advise me to do so at this point in the proceedings or rather do it in a separate PR once this one could be merged?

Reply:
I think a separate PR is fine. The problem is that a parse failure here silently 0s out the number. I'm not sure that's what we want.